### PR TITLE
[AArch64][GlobalISel] Legalization for small anyext/sext/zext

### PR DESCRIPTION
**Pull request by davemgreen**
Similar to #85625, some of the codegen is still far from optimal but this helps fix quite a few fallback cases.

### CODE REVIEW

**Comment by AtariDreams** [comment]
> Similar to #85625, some of the codegen is still far from optimal but this helps fix quite a few fallback cases.

quite a few regressions here. perhaps refine that first

**Review by AZero13 — CHANGES_REQUESTED** [review]
quite a few regressions here. perhaps refine that first

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 1]
```diff
@@ -611,7 +611,9 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
                Query.Types[0].isVector() &&
                (Query.Types[1].getScalarSizeInBits() == 8 ||
                 Query.Types[1].getScalarSizeInBits() == 16);
-      });
+      })
+      .clampMinNumElements(1, s8, 8)
+      .clampMinNumElements(1, s16, 4);
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:616]: Should we fold these into the MaxNumElements rules above, using `clampNumElements()`?

**Review by aemerson — COMMENTED** [review]
AFAICT the regressions are just a result of us being able to successfully compile instead of falling back to SDAG. In which case we should optimize those later. Not falling back is more important.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 1 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:616]: I think the idea was (it was for trunc at least), that we clamp to the largest size above, split using lower if the input size is > 2* output size, and then clamp the min size. I'm not sure this is perfect at the mo, but it hopefully stops us from jumping back and forth between smaller and larger vector sizes.

**Review by davemgreen — COMMENTED** [review]
> AFAICT the regressions are just a result of us being able to successfully compile instead of falling back to SDAG. In which case we should optimize those later. Not falling back is more important.

Yeah that sounds good, and matches what we've been doing so far. I think we are quite close for a lot of operations, and we can start trying to improve the more obvious problems with the codegen.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 1 above)
Comment (aemerson) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:616]: SGTM. At some point we might want to think about introducing some meta-rules for higher level expressions of legalization strategies.

**Review by aemerson — APPROVED** [review]

**Pull request merged**